src/weather.cpp

```diff
@@ -175,37 +166,71 @@ void Weather::DrawSnow(Bitmap& dst) {
 
 void Weather::DrawFog(Bitmap& dst) {
 	if (!fog_bitmap) {
-		fog_bitmap = Bitmap::Create(fog_image, sizeof(fog_image));
-		if (tone_effect != Tone()) {
-			fog_bitmap->ToneBlit(0, 0, *fog_bitmap, fog_bitmap->GetRect(), tone_effect, Opacity::opaque, true);
-		}
+		CreateFogOverlay();
 	}
 
-	DrawFogOverlay(dst, *fog_bitmap);
+	DrawFogOverlay(dst, *fog_bitmap, fog_tone_bitmap);
 }
 
 void Weather::DrawSandstorm(Bitmap& dst) {
 	if (!sand_bitmap) {
-		sand_bitmap = Bitmap::Create(sand_image, sizeof(sand_image));
-
-		if (tone_effect != Tone()) {
-			sand_bitmap->ToneBlit(0, 0, *sand_bitmap, sand_bitmap->GetRect(), tone_effect, Opacity::opaque, true);
-		}
+		CreateFogOverlay();
 	}
 
-	DrawFogOverlay(dst, *sand_bitmap);
+	DrawFogOverlay(dst, *sand_bitmap, sand_tone_bitmap);
 
 	// FIXME: Figure out sand particules
 	// 4 colors: white, red, orange, yellow, 1 pixel wide, 2 pixels tall.
 }
 
-void Weather::DrawFogOverlay(Bitmap& dst, const Bitmap& src) {
+void Weather::CreateFogOverlay() {
+	uint32_t fog_pixels[fog_overlay_num_colors];
+	uint32_t sand_pixels[fog_overlay_num_colors];
+
+	for (int i = 0; i < fog_overlay_num_colors; ++i) {
+		auto fc = fog_overlay_colors[i];
+		auto sc = sand_overlay_colors[i];
+		fog_pixels[i] = Bitmap::pixel_format.rgba_to_uint32_t(fc.red, fc.green, fc.blue, fc.alpha);
+		sand_pixels[i] = Bitmap::pixel_format.rgba_to_uint32_t(sc.red, sc.green, sc.blue, sc.alpha);
+	}
+
+	const auto h = fog_overlay_tile_height;
+	const auto w = fog_overlay_tile_width;
+
+	fog_bitmap = Bitmap::Create(w, h);
+	sand_bitmap = Bitmap::Create(w, h);
+
+	auto* fog_img = reinterpret_cast<uint32_t*>(fog_bitmap->pixels());
+	auto* sand_img = reinterpret_cast<uint32_t*>(sand_bitmap->pixels());
+
+	for (int i = 0; i < w * h; ++i) {
+		// FIMXE: How well does this match RPG_RT textures?
+		int px = Utils::GetRandomNumber(0, fog_overlay_num_colors - 1);
+		fog_img[i] = fog_pixels[px];
+		sand_img[i] = sand_pixels[px];
```

Comment:
@Ghabry do we need to support non 32bit rgba pixel formats? This one assumes each pixel fits into a uint32_t.

Reply:
This is fine.
16bit support is nice to have (memory usage) but due to how pixman works this is a performance killer, so currently all tested code paths are 32bit only.

Reply:
I added a fixme note, in case we do decide to get more fancy with pixel formats later on.